File: src/session/compacter.ts
import { countMessagesTokens, getCompactThreshold as resolveThreshold } from "../common/tokenizer";

export type CompactionDecision = {
  shouldCompact: boolean;
  estimatedTokens: number;
  compactUpToIndex?: number;
  keepFromIndex?: number;
};

export function shouldCompactContext(options: {
  messages: Array<{ role: string; content: string | unknown; compacted?: boolean }>;
  model: string;
  threshold?: number;
}): CompactionDecision {
  const threshold = options.threshold ?? resolveThreshold(options.model);
  const activeMessages = options.messages.filter((m) => !m.compacted);
  const estimatedTokens = countMessagesTokens(activeMessages);

  if (estimatedTokens < threshold) {
    return { shouldCompact: false, estimatedTokens };
  }

  // Phase boundary heuristic: look for recent successful tool results or system prompts
  // Instead of an arbitrary 30%, we accumulate tokens from the end to ensure we drop below the threshold
  // We want the kept messages to be roughly 60-70% of the threshold at most.
  const targetKeepTokens = Math.floor(threshold * 0.6);
  let keptTokens = 0;
  let boundaryIndex = 0;

  for (let i = activeMessages.length - 1; i >= 0; i--) {
    const msg = activeMessages[i];
    const content = typeof msg.content === "string" ? msg.content : JSON.stringify(msg.content ?? "");
    // Approximate token count to avoid expensive full tokenization in a tight loop
    // 1 char ~ 0.25 tokens for English/Code, 0.55 for CJK.
    const isMultibyte = /[^\x00-\x7F]/.test(content);
    const msgTokens = Math.ceil(content.length * (isMultibyte ? 0.55 : 0.32));

    if (keptTokens + msgTokens > targetKeepTokens) {
      boundaryIndex = i + 1;
      break;
    }
    keptTokens += msgTokens;
  }

  // Fallback to ensure we compact AT LEAST 1 message if we are above threshold,
  // but also try not to compact the very last message if possible.
  boundaryIndex = Math.min(Math.max(1, boundaryIndex), activeMessages.length - 1);

  // Try to find a clean break (user or system message) near the boundary
  for (let i = boundaryIndex; i < activeMessages.length - 1; i++) {
    const msg = activeMessages[i];
    if (msg.role === "user" || msg.role === "system") {
      boundaryIndex = i;
      break;
    }
  }

  const compactUpToIndex = boundaryIndex;

  const originalIndices: number[] = [];
  for (let i = 0; i < options.messages.length; i++) {
    if (!options.messages[i].compacted) {
      originalIndices.push(i);
    }
  }

  const keepFromIndex = originalIndices[compactUpToIndex] ?? options.messages.length;

  return {
    shouldCompact: true,
    estimatedTokens,
    compactUpToIndex:
      originalIndices[compactUpToIndex - 1] ?? options.messages.length - (activeMessages.length - boundaryIndex),
    keepFromIndex,
  };
}

File: internal/tokenizer/compacter.go
package tokenizer

import (
	"unicode"
)

type Message struct {
	Role      string
	Content   string
	Compacted bool
}

type CompactionDecision struct {
	ShouldCompact    bool
	EstimatedTokens  int
	CompactUpToIndex int
	KeepFromIndex    int
}

// EstimateTokens calculates tokens with adjustments for different character types.
// Uses empirical ratios: ~0.25 tokens/char for ASCII, ~0.35 for extended Latin,
// ~0.5 for numbers/symbols, ~1.5 for CJK, ~0.1 for whitespace.
func EstimateTokens(content string) int {
	var total float64
	for _, r := range content {
		switch {
		case r == ' ' || r == '\t' || r == '\n' || r == '\r':
			total += 0.1
		case r >= '0' && r <= '9':
			total += 0.5
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z':
			total += 0.25
		case r >= 0xC0 && r <= 0x024F: // Extended Latin
			total += 0.35
		case unicode.Is(unicode.Han, r) || unicode.In(r, unicode.Hiragana, unicode.Katakana, unicode.Hangul):
			total += 1.5
		default:
			total += 0.45
		}
	}
	if total < 1 {
		total = 1
	}
	return int(total)
}

func ShouldCompactContext(messages []Message, threshold int) CompactionDecision {
	var activeMessages []Message
	var originalIndices []int
	for idx, m := range messages {
		if !m.Compacted {
			activeMessages = append(activeMessages, m)
			originalIndices = append(originalIndices, idx)
		}
	}

	var totalTokens int
	for _, m := range activeMessages {
		totalTokens += EstimateTokens(m.Content)
	}

	if totalTokens < threshold {
		return CompactionDecision{ShouldCompact: false, EstimatedTokens: totalTokens}
	}

	targetKeepTokens := int(float64(threshold) * 0.6)
	keptTokens := 0
	boundaryIndex := 0

	for i := len(activeMessages) - 1; i >= 0; i-- {
		msgTokens := EstimateTokens(activeMessages[i].Content)
		if keptTokens+msgTokens > targetKeepTokens {
			boundaryIndex = i + 1
			break
		}
		keptTokens += msgTokens
	}

	if boundaryIndex < 1 {
		boundaryIndex = 1
	}
	if boundaryIndex >= len(activeMessages) {
		boundaryIndex = len(activeMessages) - 1
	}

	for i := boundaryIndex; i < len(activeMessages)-1; i++ {
		msg := activeMessages[i]
		if msg.Role == "user" || msg.Role == "system" {
			boundaryIndex = i
			break
		}
	}

	return CompactionDecision{
		ShouldCompact:    true,
		EstimatedTokens:  totalTokens,
		CompactUpToIndex: originalIndices[boundaryIndex-1],
		KeepFromIndex:    originalIndices[boundaryIndex],
	}
}
